File: clients/node/src/games/chess.js
'use strict';
const { BaseGame } = require('./base');
const { t } = require('../i18n');

class Chess extends BaseGame {
  constructor() {
    super();
    this.board = {};
    const back = ['R','N','B','Q','K','B','N','R'];
    for (let c = 0; c < 8; c++) {
      const f = String.fromCharCode(97 + c);
      this.board[`${f}1`] = ['w', back[c]];
      this.board[`${f}2`] = ['w', 'P'];
      this.board[`${f}7`] = ['b', 'P'];
      this.board[`${f}8`] = ['b', back[c]];
    }
    this.players = [];
    this._cmap = {};
    this._turnIdx = 0;
    this._castle = { w: { K: true, Q: true }, b: { K: true, Q: true } };
    this._enPassant = null;
    this._over = false;
    this._winner = null;
  }

  start(players) {
    this.players = players;
    this._cmap = { [players[0]]: 'w', [players[1]]: 'b' };
  }

  currentTurn() {
    return this.players[this._turnIdx] || null;
  }

  _row(sq) { return parseInt(sq[1]) - 1; }
  _col(sq) { return sq.charCodeAt(0) - 97; }
  _sq(r, c) { return String.fromCharCode(97 + c) + (r + 1); }
  _ok(r, c) { return r >= 0 && r < 8 && c >= 0 && c < 8; }
  _color(sq) { const p = this.board[sq]; return p ? p[0] : null; }

  _attacked(sq, by) {
    const r = this._row(sq), c = this._col(sq);
    const pd = by === 'w' ? -1 : 1;
    for (const dc of [-1, 1]) {
      const pr = r + pd, pc = c + dc;
      if (this._ok(pr, pc)) {
        const s = this._sq(pr, pc);
        if (this.board[s] && this.board[s][0] === by && this.board[s][1] === 'P') return true;
      }
    }
    for (const [dr, dc] of [[-2,-1],[-2,1],[-1,-2],[-1,2],[1,-2],[1,2],[2,-1],[2,1]]) {
      if (this._ok(r+dr, c+dc)) {
        const s = this._sq(r+dr, c+dc);
        if (this.board[s] && this.board[s][0] === by && this.board[s][1] === 'N') return true;
      }
    }
    for (let dr = -1; dr <= 1; dr++) {
      for (let dc = -1; dc <= 1; dc++) {
        if (dr === 0 && dc === 0) continue;
        if (this._ok(r+dr, c+dc)) {
          const s = this._sq(r+dr, c+dc);
          if (this.board[s] && this.board[s][0] === by && this.board[s][1] === 'K') return true;
        }
      }
    }
    for (const [dirs, pts] of [
      [[[0,1],[0,-1],[1,0],[-1,0]], new Set(['R','Q'])],
      [[[1,1],[1,-1],[-1,1],[-1,-1]], new Set(['B','Q'])],
    ]) {
      for (const [dr, dc] of dirs) {
        let nr = r + dr, nc = c + dc;
        while (this._ok(nr, nc)) {
          const s = this._sq(nr, nc);
          if (this.board[s]) {
            if (this.board[s][0] === by && pts.has(this.board[s][1])) return true;
            break;
          }
          nr += dr; nc += dc;
        }
      }
    }
    return false;
  }

  _kingSq(color) {
    for (const [s, p] of Object.entries(this.board)) {
      if (p[0] === color && p[1] === 'K') return s;
    }
    return null;
  }

  _inCheck(color) {
    const ks = this._kingSq(color);
    const opp = color === 'w' ? 'b' : 'w';
    return ks ? this._attacked(ks, opp) : false;
  }

  _pseudoTargets(frm) {
    if (!this.board[frm]) return [];
    const [color, piece] = this.board[frm];
    const r = this._row(frm), c = this._col(frm);
    const res = [];
    if (piece === 'P') {
      const d = color === 'w' ? 1 : -1;
      const fwd = this._sq(r + d, c);
      if (this._ok(r + d, c) && !this.board[fwd]) {
        res.push(fwd);
        const start = color === 'w' ? 1 : 6;
        const fwd2 = this._sq(r + 2*d, c);
        if (r === start && !this.board[fwd2]) res.push(fwd2);
      }
      for (const dc of [-1, 1]) {
        if (this._ok(r + d, c + dc)) {
          const s = this._sq(r + d, c + dc);
          if ((this._color(s) !== null && this._color(s) !== color) || s === this._enPassant) {
            res.push(s);
          }
        }
      }
    } else if (piece === 'N') {
      for (const [dr, dc] of [[-2,-1],[-2,1],[-1,-2],[-1,2],[1,-2],[1,2],[2,-1],[2,1]]) {
        if (this._ok(r+dr, c+dc)) {
          const s = this._sq(r+dr, c+dc);
          if (this._color(s) !== color) res.push(s);
        }
      }
    } else if (piece === 'B' || piece === 'R' || piece === 'Q') {
      const dirs = [];
      if (piece === 'B' || piece === 'Q') dirs.push(...[[1,1],[1,-1],[-1,1],[-1,-1]]);
      if (piece === 'R' || piece === 'Q') dirs.push(...[[0,1],[0,-1],[1,0],[-1,0]]);
      for (const [dr, dc] of dirs) {
        let nr = r+dr, nc = c+dc;
        while (this._ok(nr, nc)) {
          const s = this._sq(nr, nc);
          if (this._color(s) === color) break;
          res.push(s);
          if (this.board[s]) break;
          nr += dr; nc += dc;
        }
      }
    } else if (piece === 'K') {
      for (let dr = -1; dr <= 1; dr++) {
        for (let dc = -1; dc <= 1; dc++) {
          if (dr === 0 && dc === 0) continue;
          if (this._ok(r+dr, c+dc)) {
            const s = this._sq(r+dr, c+dc);
            if (this._color(s) !== color) res.push(s);
          }
        }
      }
      const rn = color === 'w' ? 1 : 8;
      const opp = color === 'w' ? 'b' : 'w';
      if (this._castle[color].K && !this._inCheck(color)) {
        const f = `f${rn}`, g = `g${rn}`;
        if (!this.board[f] && !this.board[g] && !this._attacked(f, opp) && !this._attacked(g, opp)) {
          res.push(g);
        }
      }
      if (this._castle[color].Q && !this._inCheck(color)) {
        const b2 = `b${rn}`, c2 = `c${rn}`, d2 = `d${rn}`;
        if (!this.board[b2] && !this.board[c2] && !this.board[d2] &&
            !this._attacked(c2, opp) && !this._attacked(d2, opp)) {
          res.push(c2);
        }
      }
    }
    return res;
  }

  _push(frm, to, promo = 'Q') {
    const saved = {
      board: { ...this.board },
      castle: { w: { ...this._castle.w }, b: { ...this._castle.b } },
      ep: this._enPassant,
    };
    const [color, piece] = this.board[frm];
    delete this.board[frm];
    this._enPassant = null;
    if (piece === 'P' && to === saved.ep) {
      const capR = this._row(to) + (color === 'w' ? -1 : 1);
      delete this.board[this._sq(capR, this._col(to))];
    }
    if (piece === 'P' && Math.abs(this._row(to) - this._row(frm)) === 2) {
      this._enPassant = this._sq(Math.floor((this._row(frm) + this._row(to)) / 2), this._col(frm));
    }
    if (piece === 'K') {
      this._castle[color] = { K: false, Q: false };
      const rn = this._row(frm) + 1;
      const dc = this._col(to) - this._col(frm);
      if (dc === 2) { this.board[`f${rn}`] = this.board[`h${rn}`]; delete this.board[`h${rn}`]; }
      else if (dc === -2) { this.board[`d${rn}`] = this.board[`a${rn}`]; delete this.board[`a${rn}`]; }
    }
    if (piece === 'R') {
      const rn = color === 'w' ? 1 : 8;
      if (frm === `a${rn}`) this._castle[color].Q = false;
      if (frm === `h${rn}`) this._castle[color].K = false;
    }
    if (piece === 'P' && (this._row(to) === 7 || this._row(to) === 0)) {
      this.board[to] = [color, promo];
    } else {
      this.board[to] = [color, piece];
    }
    return saved;
  }

  _pop(saved) {
    this.board = saved.board;
    this._castle = saved.castle;
    this._enPassant = saved.ep;
  }

  _legal(frm, to, promo = 'Q') {
    if (!this._pseudoTargets(frm).includes(to)) return false;
    const color = this.board[frm][0];
    const s = this._push(frm, to, promo);
    const ok = !this._inCheck(color);
    this._pop(s);
    return ok;
  }

  _allLegal(color) {
    const moves = [];
    for (const frm of Object.keys(this.board)) {
      if (this.board[frm] && this.board[frm][0] === color) {
        for (const to of this._pseudoTargets(frm)) {
          if (this._legal(frm, to)) moves.push([frm, to]);
        }
      }
    }
    return moves;
  }

  validateMove(playerId, moveData) {
    if (this._over || playerId !== this.currentTurn()) return false;
    const { from: frm, to } = moveData;
    if (!frm || !to || frm.length !== 2 || to.length !== 2) return false;
    if (!this.board[frm]) return false;
    if (this.board[frm][0] !== this._cmap[playerId]) return false;
    return this._legal(frm, to, moveData.promotion || 'Q');
  }

  applyMove(playerId, moveData) {
    this._push(moveData.from, moveData.to, moveData.promotion || 'Q');
    this._turnIdx = 1 - this._turnIdx;
    const nxtColor = this._cmap[this.players[this._turnIdx]];
    if (!this._allLegal(nxtColor).length) {
      this._over = true;
      this._winner = this._inCheck(nxtColor) ? this.players[1 - this._turnIdx] : null;
    }
  }

  isOver() {
    return [this._over, this._winner];
  }

  render(perspective) {
    const lines = [t('chess.title'), t('chess.board_header')];
    for (let rank = 7; rank >= 0; rank--) {
      let row = `${rank + 1} `;
      for (let file = 0; file < 8; file++) {
        const sq = this._sq(rank, file);
        const p = this.board[sq];
        if (p) row += (p[0] === 'w' ? p[1] : p[1].toLowerCase()) + ' ';
        else row += '. ';
      }
      lines.push(row);
    }
    lines.push(t('chess.turn', { player: this.currentTurn() }));
    return lines.join('\n');
  }

  getState(perspective) {
    return {
      board: Object.fromEntries(Object.entries(this.board).map(([s, p]) => [s, [...p]])),
      turn: this.currentTurn(),
      players: this.players,
      check: !this._over ? this._inCheck(this._cmap[this.currentTurn()] || 'w') : false,
    };
  }
}

module.exports = { Chess };

File: clients/python/games/chess.py
from typing import List, Optional
from games.base import BaseGame
from framework.i18n import t


class Chess(BaseGame):
    min_players = max_players = 2

    def __init__(self):
        self.board = {
            'a1': ('w', 'R'), 'b1': ('w', 'N'), 'c1': ('w', 'B'), 'd1': ('w', 'Q'),
            'e1': ('w', 'K'), 'f1': ('w', 'B'), 'g1': ('w', 'N'), 'h1': ('w', 'R'),
            **{f'{f}2': ('w', 'P') for f in 'abcdefgh'},
            **{f'{f}7': ('b', 'P') for f in 'abcdefgh'},
            'a8': ('b', 'R'), 'b8': ('b', 'N'), 'c8': ('b', 'B'), 'd8': ('b', 'Q'),
            'e8': ('b', 'K'), 'f8': ('b', 'B'), 'g8': ('b', 'N'), 'h8': ('b', 'R'),
        }
        self.players = []
        self._cmap = {}
        self._turn_idx = 0
        self._castle = {'w': {'K': True, 'Q': True}, 'b': {'K': True, 'Q': True}}
        self._en_passant = None
        self._over = False
        self._winner = None

    def start(self, players):
        self.players = players
        self._cmap = {players[0]: 'w', players[1]: 'b'}

    def current_turn(self):
        return self.players[self._turn_idx] if self.players else None

    def _row(self, sq): return int(sq[1]) - 1
    def _col(self, sq): return ord(sq[0]) - ord('a')
    def _sq(self, r, c): return chr(ord('a') + c) + str(r + 1)
    def _ok(self, r, c): return 0 <= r < 8 and 0 <= c < 8

    def _color(self, sq):
        p = self.board.get(sq)
        return p[0] if p else None

    def _attacked(self, sq, by):
        r, c = self._row(sq), self._col(sq)
        pd = -1 if by == 'w' else 1
        for dc in [-1, 1]:
            pr, pc = r + pd, c + dc
            if self._ok(pr, pc) and self.board.get(self._sq(pr, pc)) == (by, 'P'):
                return True
        for dr, dc in [(-2, -1), (-2, 1), (-1, -2), (-1, 2), (1, -2), (1, 2), (2, -1), (2, 1)]:
            s = self._sq(r + dr, c + dc) if self._ok(r + dr, c + dc) else None
            if s and self.board.get(s) == (by, 'N'):
                return True
        for dr in [-1, 0, 1]:
            for dc in [-1, 0, 1]:
                if dr == dc == 0:
                    continue
                s = self._sq(r + dr, c + dc) if self._ok(r + dr, c + dc) else None
                if s and self.board.get(s) == (by, 'K'):
                    return True
        for dirs, pts in [
            ([(0, 1), (0, -1), (1, 0), (-1, 0)], {'R', 'Q'}),
            ([(1, 1), (1, -1), (-1, 1), (-1, -1)], {'B', 'Q'}),
        ]:
            for dr, dc in dirs:
                nr, nc = r + dr, c + dc
                while self._ok(nr, nc):
                    s = self._sq(nr, nc)
                    if s in self.board:
                        if self.board[s][0] == by and self.board[s][1] in pts:
                            return True
                        break
                    nr += dr
                    nc += dc
        return False

    def _king_sq(self, color):
        return next((s for s, p in self.board.items() if p == (color, 'K')), None)

    def _in_check(self, color):
        ks = self._king_sq(color)
        opp = 'b' if color == 'w' else 'w'
        return self._attacked(ks, opp) if ks else False

    def _pseudo_targets(self, frm):
        if frm not in self.board:
            return []
        color, piece = self.board[frm]
        r, c = self._row(frm), self._col(frm)
        res = []
        if piece == 'P':
            d = 1 if color == 'w' else -1
            fwd = self._sq(r + d, c)
            if self._ok(r + d, c) and fwd not in self.board:
                res.append(fwd)
                start = 1 if color == 'w' else 6
                fwd2 = self._sq(r + 2 * d, c)
                if r == start and fwd2 not in self.board:
                    res.append(fwd2)
            for dc in [-1, 1]:
                if self._ok(r + d, c + dc):
                    s = self._sq(r + d, c + dc)
                    if self._color(s) not in (None, color) or s == self._en_passant:
                        res.append(s)
        elif piece == 'N':
            for dr, dc in [(-2, -1), (-2, 1), (-1, -2), (-1, 2), (1, -2), (1, 2), (2, -1), (2, 1)]:
                if self._ok(r + dr, c + dc):
                    s = self._sq(r + dr, c + dc)
                    if self._color(s) != color:
                        res.append(s)
        elif piece in ('B', 'R', 'Q'):
            dirs = []
            if piece in ('B', 'Q'):
                dirs += [(1, 1), (1, -1), (-1, 1), (-1, -1)]
            if piece in ('R', 'Q'):
                dirs += [(0, 1), (0, -1), (1, 0), (-1, 0)]
            for dr, dc in dirs:
                nr, nc = r + dr, c + dc
                while self._ok(nr, nc):
                    s = self._sq(nr, nc)
                    if self._color(s) == color:
                        break
                    res.append(s)
                    if s in self.board:
                        break
                    nr += dr
                    nc += dc
        elif piece == 'K':
            for dr in [-1, 0, 1]:
                for dc in [-1, 0, 1]:
                    if dr == dc == 0:
                        continue
                    if self._ok(r + dr, c + dc):
                        s = self._sq(r + dr, c + dc)
                        if self._color(s) != color:
                            res.append(s)
            row_n = 1 if color == 'w' else 8
            opp = 'b' if color == 'w' else 'w'
            if self._castle[color]['K'] and not self._in_check(color):
                f, g = f'f{row_n}', f'g{row_n}'
                if f not in self.board and g not in self.board:
                    if not self._attacked(f, opp) and not self._attacked(g, opp):
                        res.append(g)
            if self._castle[color]['Q'] and not self._in_check(color):
                b2, c2, d2 = f'b{row_n}', f'c{row_n}', f'd{row_n}'
                if b2 not in self.board and c2 not in self.board and d2 not in self.board:
                    if not self._attacked(c2, opp) and not self._attacked(d2, opp):
                        res.append(c2)
        return res

    def _push(self, frm, to, promo='Q'):
        saved = {
            'board': dict(self.board),
            'castle': {c: dict(v) for c, v in self._castle.items()},
            'ep': self._en_passant,
        }
        color, piece = self.board.pop(frm)
        self._en_passant = None
        if piece == 'P' and to == saved['ep']:
            cap_r = self._row(to) + (-1 if color == 'w' else 1)
            self.board.pop(self._sq(cap_r, self._col(to)), None)
        if piece == 'P' and abs(self._row(to) - self._row(frm)) == 2:
            self._en_passant = self._sq(
                (self._row(frm) + self._row(to)) // 2, self._col(frm))
        if piece == 'K':
            self._castle[color] = {'K': False, 'Q': False}
            rn = self._row(frm) + 1
            if self._col(to) - self._col(frm) == 2:
                self.board[f'f{rn}'] = self.board.pop(f'h{rn}')
            elif self._col(to) - self._col(frm) == -2:
                self.board[f'd{rn}'] = self.board.pop(f'a{rn}')
        if piece == 'R':
            rn = 1 if color == 'w' else 8
            if frm == f'a{rn}':
                self._castle[color]['Q'] = False
            if frm == f'h{rn}':
                self._castle[color]['K'] = False
        if piece == 'P' and (self._row(to) == 7 or self._row(to) == 0):
            self.board[to] = (color, promo)
        else:
            self.board[to] = (color, piece)
        return saved

    def _pop(self, saved):
        self.board = saved['board']
        self._castle = saved['castle']
        self._en_passant = saved['ep']

    def _legal(self, frm, to, promo='Q'):
        if to not in self._pseudo_targets(frm):
            return False
        color = self.board[frm][0]
        s = self._push(frm, to, promo)
        ok = not self._in_check(color)
        self._pop(s)
        return ok

    def _all_legal(self, color):
        return [(f, t) for f in list(self.board)
                if self.board.get(f, (None,))[0] == color
                for t in self._pseudo_targets(f) if self._legal(f, t)]

    def validate_move(self, player_id, move_data):
        if self._over or player_id != self.current_turn():
            return False
        frm, to = move_data.get('from', ''), move_data.get('to', '')
        if len(frm) != 2 or len(to) != 2:
            return False
        if frm not in self.board:
            return False
        if self.board[frm][0] != self._cmap.get(player_id):
            return False
        return self._legal(frm, to, move_data.get('promotion', 'Q'))

    def apply_move(self, player_id, move_data):
        self._push(move_data['from'], move_data['to'], move_data.get('promotion', 'Q'))
        self._turn_idx = 1 - self._turn_idx
        nxt_color = self._cmap[self.players[self._turn_idx]]
        if not self._all_legal(nxt_color):
            self._over = True
            self._winner = (self.players[1 - self._turn_idx]
                            if self._in_check(nxt_color) else None)

    def is_over(self):
        return self._over, self._winner

    def render(self, perspective=None):
        lines = [t('chess.title'), t('chess.board_header')]
        for rank in range(7, -1, -1):
            row = str(rank + 1) + ' '
            for file in range(8):
                sq = self._sq(rank, file)
                p = self.board.get(sq)
                if p:
                    row += (p[1] if p[0] == 'w' else p[1].lower()) + ' '
                else:
                    row += '. '
            lines.append(row)
        lines.append(t('chess.turn', player=self.current_turn()))
        return '\n'.join(lines)

    def get_state(self, perspective: Optional[str] = None) -> dict:
        return {
            'board': {s: list(p) for s, p in self.board.items()},
            'turn': self.current_turn(),
            'players': self.players,
            'castle': self._castle,
            'en_passant': self._en_passant,
            'check': (self._in_check(self._cmap.get(self.current_turn(), 'w'))
                      if not self._over else False),
        }

    def load_state(self, data: dict, perspective: Optional[str] = None) -> None:
        if not data:
            return
        if 'board' in data:
            self.board = {sq: tuple(p) for sq, p in data['board'].items()}
        if 'players' in data:
            self.players = list(data['players'])
            if len(self.players) == 2:
                self._cmap = {self.players[0]: 'w', self.players[1]: 'b'}
        if 'turn' in data and data['turn'] in self.players:
            self._turn_idx = self.players.index(data['turn'])
        if 'castle' in data:
            self._castle = data['castle']
        if 'en_passant' in data:
            self._en_passant = data['en_passant']

    def parse_input(self, raw: str) -> Optional[dict]:
        import json as _json
        raw = raw.strip()
        if raw.startswith('{'):
            try:
                obj = _json.loads(raw)
                if isinstance(obj, dict):
                    return obj
            except ValueError:
                pass
        parts = raw.split()
        if len(parts) == 2:
            return {'from': parts[0], 'to': parts[1]}
        return None

    def get_help(self) -> List[str]:
        return [
            'Standard chess. Castling, en passant, and promotion all supported.',
            'Move: <from> <to>   e.g. "e2 e4"   castle: "e1 g1"',
        ]
